### PR TITLE
add password hashing

### LINKED ISSUE #2: Store user passwords securely
In the initial minimal implementation, passwords are just stored as plain text.
These need to be stored more securely as an example of "good" practices.

### PR DESCRIPTION
**Pull request by Giesch**
* resolves #2
* avoids logging or returning passwords
* adds some custom error types for the app/web boundary

I'd love to get feedback on this. I feel like there's probably a better way to do error handling, but I don't know the ecosystem well enough. I'm also not sure if the 'into PublicUser' conversion really belongs in the app or web layer.

### CODE REVIEW

**Comment by colinbankier** [comment]
Thanks for your contribution! I'll take a more detailed look over the next few days.

**Comment by colinbankier** [comment]
Thanks @Giesch - master has moved on a bit - would you like to rebase your changes on master so it can be merged in?
In terms of error handing - there are many crates for handling this but just using enums as you've done is perfectly fine until it becomes obvious something else is needed.

**Comment by Giesch** [comment]
This resolves the merge conflicts and removes the 'public user' stuff that was basically duplicated.

File: src/conduit/users.rs [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
+use diesel::pg::PgConnection;
 use diesel::prelude::*;
-use diesel::result::Error;
+use diesel::result::Error as DieselError;
 use uuid::Uuid;
 
-pub fn insert(repo: &Repo, user: NewUser) -> Result<User, Error> {
-    repo.run(move |conn| {
-        // TODO: store password not in plain text, later
+type Repo = crate::db::Repo<PgConnection>;
```
Comment (LukeMathWalker) [inline src/conduit/users.rs]: It would be better to import the type alias, instead of redefining it.

Comment (Giesch) [reply src/conduit/users.rs]: fixed with https://github.com/colinbankier/realworld-tide/pull/17/commits/d73c809d88a7b9525ce9c691508652fdd685a0b2

**Comment by colinbankier** [comment]
Great, thanks for your contribution @Giesch :+1:

**Pull request merged**